### PR TITLE
Expander refactoring [property tracking analyzers prerequisity]

### PR DESCRIPTION
**Pull request by JanKrivanek**
Contributes to #9883 (spinoff from #10009 - focusing on parts not related to BuildCheck)

### Note

There is no new functionality in this PR - just refactoring and code moving. It is required for the followup PR(s) for property tracking analyzers

### Context
Expander is using `LoggingContext` in some of the execution paths, but in it usually has a default value of `null` - so we end up with literaly hundrets of calling paths where we cannot be sure whether `LoggingContext` is available or not. The `LoggingContext` is appealing for a data pub-sub, so let's make sure it's available throughout the `Expander` calls, while at the same let's simplify the code a bit

### Changes Made
 * Added constructors to `Expander` to pass `LoggingContext` wherever possible (the only exception is the Expander usages from the ToolsetReader - where we do not know the build context yet).
 * Removed explicit LoggingContext arg from the Expander methods
  Reason: Went through all the methods in Expander and all the possible caller chains (there were many hunderts of them) - and found out there is just a single case where LoggingContext passed to expander calls differ from the LoggingContext used to construct the Expander
 * Added `ItemBucket.UpdateLoggingContext` 
    Reason:  The single differing case was in `TargetEntry.ExecuteTarget` where `ProjectLoggingContext` was used to construct the `Expander`, but then `TargetLogginContext` was used to transitively call the methods. So the updating of the context was used to bridge this case
 * Moved the conditioning of `PropertyTrackingEvaluatorDataWrapper` into the type - so that in the folowup PR we can use the type to feed data to analyzers even if proprty tracking logging is not requested
 * Renamed `UsedUninitializedProperties` -> `PropertiesUsageTracker`
 * Added more info to calls of `PropertiesUsageTracker`
   * Info about context in which the property was expanded
   * Moved the code to make this class responsible for deciding about uninitialized reads

### Testing
 Refactored and resued existing tests (as there is no new functionality)

### CODE REVIEW

File: src/Build/BackEnd/Components/RequestBuilder/ItemBucket.cs [hunk 1]
```diff
@@ -95,11 +98,20 @@ private ItemBucket()
...
         }
 
+        /// <summary>
+        /// Updates the logging context that this bucket is going to use.
+        /// </summary>
+        /// <param name="loggingContext"></param>
+        internal void UpdateLoggingContext(LoggingContext loggingContext)
+        {
+            _expander = _expander.WithLoggingContext(loggingContext);
+        }
```
Comment (ladipro) [inline src/Build/BackEnd/Components/RequestBuilder/ItemBucket.cs:113]: This is unfortunate. Not a big deal and not a blocking comment but it does reduce the cleanliness, IMO. Do instances of this class really use the logging context they are constructed with, before it's replaced here? If the answer is no, can we delay creation of these objects until the final logging context is known?

File: src/Build/Evaluation/Evaluator.cs [hunk 2]
```diff
@@ -223,12 +224,8 @@ internal class Evaluator<P, I, M, D>
...
                 string.IsNullOrEmpty(projectRootElement.ProjectFileLocation.File) ? "(null)" : projectRootElement.ProjectFileLocation.File);
 
-            // If someone sets the 'MsBuildLogPropertyTracking' environment variable to a non-zero value, wrap property accesses for event reporting.
-            if (Traits.Instance.LogPropertyTracking > 0)
-            {
-                // Wrap the IEvaluatorData<> object passed in.
-                data = new PropertyTrackingEvaluatorDataWrapper<P, I, M, D>(data, _evaluationLoggingContext, Traits.Instance.LogPropertyTracking);
-            }
+            // Wrap the IEvaluatorData<> object passed in.
+            data = new PropertyTrackingEvaluatorDataWrapper<P, I, M, D>(data, _evaluationLoggingContext, Traits.Instance.LogPropertyTracking);
```
Comment (ladipro) [inline src/Build/Evaluation/Evaluator.cs:228]: Is the unconditional wrapping expected to have negligible perf impact?

**Review by ladipro — COMMENTED** [review]
Thank you, this is good stuff. Leaving a few comments inline.

File: src/Build/Evaluation/Evaluator.cs (hunk 2 above)
Comment (JanKrivanek) [reply src/Build/Evaluation/Evaluator.cs:228]: Let's test with PerfStart

File: src/Build/BackEnd/Components/RequestBuilder/ItemBucket.cs (hunk 1 above)
Comment (JanKrivanek) [reply src/Build/BackEnd/Components/RequestBuilder/ItemBucket.cs:113]: Unfortunately if we batch targets - then buckets are prepared prior the target logging context is created (as that creation logs info about starting run of a target)

Comment (ladipro) [reply src/Build/BackEnd/Components/RequestBuilder/ItemBucket.cs:113]: Gotcha. I took a quick look and I am still wondering if there is an easy way to make it more elegant (and a bit more performant). Currently:

1. `ItemBucket` is created, internally it creates a new `Expander` with the original logging context.
2. (The original logging context is not used, the expander is not used.)
3. `UpdateLoggingContext` is called to replace the expander with a newly created one.

What if, instead, `ItemBucket` construction was split into the real constructor and an `Initialize` method. Only the latter would create the expander with the right logging context. This way there would be no throw-away expander, and the pattern of ctor+Initialize would look slightly more conventional.

Comment (JanKrivanek) [reply src/Build/BackEnd/Components/RequestBuilder/ItemBucket.cs:113]: Looks better then re-creation pattern - thx! - will try that out

File: src/Build/Evaluation/Expander.cs [hunk 3]
```diff
@@ -5493,13 +5502,57 @@ private object LateBindExecute(Exception ex, BindingFlags bindingFlags, object o
...
+        {
+            if (isArtificial)
+            {
+                return;
+            }
+
+            if (!isUninitialized)
+            {
+                return;
+            }
```
Comment (ladipro) [inline src/Build/Evaluation/Expander.cs:5533]: nit: This feels slightly unusual. Why not make the call only `if (isUninitialized && !isArtificial)` ?

File: src/Build/Evaluation/Expander.cs [hunk 4]
```diff
@@ -1517,21 +1540,26 @@ internal static string ConvertToString(object valueToConvert)
...
-            private static object LookupProperty(IPropertyProvider<T> properties, string propertyName, int startIndex, int endIndex, IElementLocation elementLocation, UsedUninitializedProperties usedUninitializedProperties, LoggingContext loggingContext = null)
+            private static object LookupProperty(IPropertyProvider<T> properties, string propertyName, int startIndex, int endIndex, IElementLocation elementLocation, PropertiesUsageTracker propertiesUsageTracker)
             {
                 T property = properties.GetProperty(propertyName, startIndex, endIndex);
 
                 object propertyValue;
 
-                if (property == null && ((endIndex - startIndex) >= 7) && MSBuildNameIgnoreCaseComparer.Default.Equals("MSBuild", propertyName, startIndex, 7))
+                bool isArtifical = property == null && ((endIndex - startIndex) >= 7) &&
+                                   MSBuildNameIgnoreCaseComparer.Default.Equals("MSBuild", propertyName, startIndex, 7);
```
Comment (ladipro) [inline src/Build/Evaluation/Expander.cs:1558]: nit: Pre-existing issue: this is just a heuristic. Do we eventually want to fix it to really detect only built-in properties using something like `ReservedPropertyNames.IsReservedProperty()` ?

**Review by ladipro — APPROVED** [review]

**Review by YuliiaKovalova — APPROVED** [review]

File: src/Build/Evaluation/Expander.cs (hunk 4 above)
Comment (JanKrivanek) [reply src/Build/Evaluation/Expander.cs:1558]: What happens if someone defines 'MSBuildXXX' property today? Is that allowed?
It feels we should rather prevent users from doing that (rather then controlling just the subset we use today).

We might actually want to include it in scope of https://github.com/dotnet/msbuild/issues/9891

Comment (ladipro) [reply src/Build/Evaluation/Expander.cs:1558]: Yep, too bad the prefix isn't reserved from the beginning. There is nothing preventing anyone from defining their own `MSBuild*` properties.

File: src/Build/Evaluation/Expander.cs [hunk 5]
```diff
@@ -5493,13 +5502,52 @@ private object LateBindExecute(Exception ex, BindingFlags bindingFlags, object o
     /// <summary>
     /// This class wraps information about properties which have been used before they are initialized.
     /// </summary>
-    internal sealed class UsedUninitializedProperties
+    internal sealed class PropertiesUsageTracker
```
[the remark this answers is not in the document]
Comment (rainersigwald) [reply src/Build/Evaluation/Expander.cs:5505]: Also I'd personally name it `PropertyUseTracker`.

File: src/Build/Instance/IElementWithLocation.cs [hunk 6]
```diff
@@ -0,0 +1,14 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Build.Construction;
+
+namespace Microsoft.Build.Execution
+{
+    internal interface IElementWithLocation
```
[the remark this answers is not in the document]
Comment (JanKrivanek) [reply src/Build/Instance/IElementWithLocation.cs:8]: Renamed to `IPropertyElementWithLocation` and xml commented

File: src/Build/Evaluation/Conditionals/NumericExpressionNode.cs [hunk 7]
```diff
@@ -23,22 +23,26 @@ internal NumericExpressionNode(string value)
...
 
-        internal override bool TryVersionEvaluate(ConditionEvaluator.IConditionEvaluationState state, out Version result, LoggingContext loggingContext = null)
+        internal override bool TryVersionEvaluate(ConditionEvaluator.IConditionEvaluationState state, out Version result)
         {
             return Version.TryParse(_value, out result);
         }
 
+        /// <inheritdoc cref="GenericExpressionNode"/>
+        internal override bool IsUnexpandedValueEmpty(ConditionEvaluator.IConditionEvaluationState state)
+            => string.IsNullOrEmpty(_value);
```
[the remark this answers is not in the document]
Comment (JanKrivanek) [reply src/Build/Evaluation/Conditionals/NumericExpressionNode.cs:44]: You're right - the only places creating Numeric tokens:

https://github.com/dotnet/msbuild/blob/393c2fea652873416c8a2028810932a4fa94403f/src/Build/Evaluation/Conditionals/Scanner.cs#L726

and

https://github.com/dotnet/msbuild/blob/393c2fea652873416c8a2028810932a4fa94403f/src/Build/Evaluation/Conditionals/Scanner.cs#L753

Both verifying there is actually some string left.


So I applied nullable checks and runtime check on not emptiness during constructions - and the method in question is now constant `false`

File: src/Build/BackEnd/Components/RequestBuilder/ItemBucket.cs [hunk 8]
```diff
@@ -95,11 +98,19 @@ private ItemBucket()
...
             _bucketSequenceNumber = bucketSequenceNumber;
         }
 
+        /// <summary>
+        /// Updates the logging context that this bucket is going to use.
+        /// </summary>
+        /// <param name="loggingContext"></param>
+        internal void Initialize(LoggingContext loggingContext)
+        {
+            _expander = new Expander<ProjectPropertyInstance, ProjectItemInstance>(_lookup, _lookup, new StringMetadataTable(_metadata), FileSystems.Default, loggingContext);
```
[the remark this answers is not in the document]
Comment (JanKrivanek) [reply src/Build/BackEnd/Components/RequestBuilder/ItemBucket.cs:111]: There is luckily just single accessing property - added check there

File: src/Build/BackEnd/Components/RequestBuilder/ItemBucket.cs [hunk 9]
```diff
@@ -58,8 +60,9 @@ internal sealed class ItemBucket : IComparable
         /// <summary>
         /// Private default constructor disallows parameterless instantiation.
```
[the remark this answers is not in the document]
Comment (JanKrivanek) [reply src/Build/BackEnd/Components/RequestBuilder/ItemBucket.cs:61]: Adjusted the comment.
(There is yet another parametrized ctor - so parameterless construction is not possible regardless of the signature of this one. It's actually used just to create dummy item for binary searching).

File: src/Build/Evaluation/Conditionals/MultipleComparisonExpressionNode.cs [hunk 10]
```diff
@@ -51,8 +51,16 @@ internal override bool BoolEvaluate(ConditionEvaluator.IConditionEvaluationState
...
             // That means we don't have to fully expand a condition like " '@(X)' == '' "
             // which is a performance advantage if @(X) is a huge item list.
-            bool leftEmpty = LeftChild.EvaluatesToEmpty(state, loggingContext);
-            bool rightEmpty = RightChild.EvaluatesToEmpty(state, loggingContext);
+
+            // this is the possible case of '$(a)' == ''
+            if (LeftChild.IsUnexpandedValueEmpty(state) ||
+                RightChild.IsUnexpandedValueEmpty(state))
+            {
+                state.PropertiesUsageTracker.PropertyReadContext = PropertyReadContext.ConditionEvaluationWithOneSideEmpty;
```
[the remark this answers is not in the document]
Comment (JanKrivanek) [reply src/Build/Evaluation/Conditionals/MultipleComparisonExpressionNode.cs:59]: Updated the comment to:

```
            // This is the possible case of an expression similar to '$(a)' == '', where a usage of uninitialized
            //  property is reasonable and should not be flagged by uninitialized reads detection.
            // So if at least one side is empty, we know to signal to PropertiesUseTracker to not flag in this scope.
            // The other side might not be property at all - that's fine, as then PropertiesUseTracker won't be even called.
```

Please let me know if it makes sense.

File: src/Build/Evaluation/Expander.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (JanKrivanek) [reply src/Build/Evaluation/Expander.cs:5505]: 👍 to both suggestions! - done :-)

File: src/Build/BackEnd/Components/RequestBuilder/IntrinsicTasks/PropertyGroupIntrinsicTask.cs [hunk 11]
```diff
@@ -97,6 +101,7 @@ internal override void ExecuteTask(Lookup lookup)
                         foreach (ItemBucket bucket in buckets)
                         {
                             bucket.LeaveScope();
+                            bucket.Expander.PropertiesUsageTracker.ResetPropertyReadContext(false);
```
[the remark this answers is not in the document]
Comment (JanKrivanek) [reply src/Build/BackEnd/Components/RequestBuilder/IntrinsicTasks/PropertyGroupIntrinsicTask.cs:104]: Added clarifying comment.

```
// We are now done processing this property - so no need to pop its previous context.
```

File: src/Build/Evaluation/Evaluator.cs (hunk 2 above)
Comment (JanKrivanek) [reply src/Build/Evaluation/Evaluator.cs:228]: PerfStar is currently down with unclear timelines. So I run a manual perf tests- numbers included here:

https://github.com/dotnet/msbuild/pull/10125

tl;dr;: Basic measurement did not reveal measurable difference
